File: src/unicode_python/segmentation/grapheme.py
"""UAX #29 default extended grapheme cluster segmentation.

A port of the Lean algorithm
``Unicode.Segmentation.GraphemeBreak.graphemeBreaks``. The active Lean tree
proves ``graphemeBreaks_eq_spec``, relating that algorithm to the declarative
UAX #29 GB1-GB999 specification. The state fields, rule order, and transition
below mirror that reference.

The property tables are grouped by property value (as in the UCD source), not
globally sorted by code point, so lookups scan linearly for the covering range,
mirroring the verified Lean ``find?``. Each class is a partition, so the first
covering range is the only one.
"""

from __future__ import annotations

from dataclasses import dataclass

from unicode_python.segmentation.grapheme_tables import (
    EXTPICT_RANGES,
    GCB_RANGES,
    INCB_RANGES,
)


def lookup_gcb(cp: int) -> str:
    """Grapheme_Cluster_Break class of ``cp``, ``"Other"`` when uncovered."""
    for lo, hi, cls in GCB_RANGES:
        if lo <= cp <= hi:
            return cls
    return "Other"


def lookup_incb(cp: int) -> str:
    """Indic_Conjunct_Break class of ``cp``, ``"None"`` when uncovered."""
    for lo, hi, cls in INCB_RANGES:
        if lo <= cp <= hi:
            return cls
    return "None"


def is_ext_pict(cp: int) -> bool:
    """Whether ``cp`` has the Extended_Pictographic property."""
    return any(lo <= cp <= hi for lo, hi in EXTPICT_RANGES)


@dataclass
class _State:
    """Running scan state, mirroring the Lean ``State``."""

    prev_class: str | None
    epic_state: str  # "none" | "after_ep" | "after_ep_zwj"  (GB11)
    incb_state: str  # "none" | "consonant" | "linker"        (GB9c)
    ri_run: int


def _initial() -> _State:
    return _State(prev_class=None, epic_state="none", incb_state="none", ri_run=0)


def _should_break_before(cp: int, s: _State) -> bool:
    """Whether a grapheme cluster break occurs immediately before ``cp``.

    Implements UAX #29 GB1-GB999 in canonical order; first match wins, the
    trailing GB999 breaks every otherwise-unmatched pair.
    """
    bc = lookup_gcb(cp)
    incb = lookup_incb(cp)
    is_ep = is_ext_pict(cp)
    pc = s.prev_class
    if pc is None:
        return True  # GB1: sot break
    if pc == "CR" and bc == "LF":
        return False  # GB3
    if pc in ("Control", "CR", "LF"):
        return True  # GB4
    if bc in ("Control", "CR", "LF"):
        return True  # GB5
    if pc == "L" and bc in ("L", "V", "LV", "LVT"):
        return False  # GB6
    if pc in ("LV", "V") and bc in ("V", "T"):
        return False  # GB7
    if pc in ("LVT", "T") and bc == "T":
        return False  # GB8
    if bc in ("Extend", "ZWJ"):
        return False  # GB9
    if bc == "SpacingMark":
        return False  # GB9a
    if pc == "Prepend":
        return False  # GB9b
    if s.incb_state == "linker" and incb == "Consonant":
        return False  # GB9c
    if s.epic_state == "after_ep_zwj" and is_ep:
        return False  # GB11
    if bc == "Regional_Indicator" and s.ri_run % 2 == 1:
        return False  # GB12/GB13
    return True  # GB999


def _advance(cp: int, s: _State) -> _State:
    """Update the running state after consuming ``cp``. Mirrors ``advance``."""
    bc = lookup_gcb(cp)
    incb = lookup_incb(cp)
    is_ep = is_ext_pict(cp)
    if is_ep:
        epic = "after_ep"
    elif s.epic_state == "after_ep" and bc == "Extend":
        epic = "after_ep"
    elif s.epic_state == "after_ep" and bc == "ZWJ":
        epic = "after_ep_zwj"
    else:
        epic = "none"
    if incb == "Consonant":
        incb_s = "consonant"
    elif s.incb_state == "consonant" and incb == "Linker":
        incb_s = "linker"
    elif s.incb_state == "consonant" and incb == "Extend":
        incb_s = "consonant"
    elif s.incb_state == "linker" and incb == "Linker":
        incb_s = "linker"
    elif s.incb_state == "linker" and incb == "Extend":
        incb_s = "linker"
    else:
        incb_s = "none"
    ri = s.ri_run + 1 if bc == "Regional_Indicator" else 0
    return _State(prev_class=bc, epic_state=epic, incb_state=incb_s, ri_run=ri)


def grapheme_breaks(cps: list[int]) -> list[bool]:
    """Boundary mask of length ``len(cps) + 1``.

    Entry ``i`` is ``True`` when a grapheme cluster break occurs immediately
    before position ``i`` -- entry ``0`` is the GB1 start-of-text break, entry
    ``len(cps)`` the GB2 end-of-text break, both always ``True``. Mirrors the
    Lean ``graphemeBreaks``.
    """
    breaks: list[bool] = []
    s = _initial()
    for cp in cps:
        breaks.append(_should_break_before(cp, s))
        s = _advance(cp, s)
    breaks.append(True)  # GB2: eot break
    return breaks


def grapheme_clusters(cps: list[int]) -> list[list[int]]:
    """Split ``cps`` into grapheme clusters (code points between boundaries)."""
    breaks = grapheme_breaks(cps)
    out: list[list[int]] = []
    cur: list[int] = []
    for i, cp in enumerate(cps):
        if breaks[i] and cur:
            out.append(cur)
            cur = []
        cur.append(cp)
    if cur:
        out.append(cur)
    return out

File: ports/dotnet/src/UnicodeSecurity/Segmentation/Grapheme.cs
// UAX #29 default extended grapheme cluster segmentation.
//
// A transcription of the Lean algorithm
// Unicode.Segmentation.GraphemeBreak.graphemeBreaks by way of
// ports/rust/src/segmentation/grapheme.rs. The active Lean tree proves
// graphemeBreaks_eq_spec, relating that algorithm to the declarative UAX #29
// GB1-GB999 specification. The state fields, rule order, and transitions below
// mirror that reference exactly.
//
// The property tables (GraphemeTables) are grouped by property value as in the
// UCD source, not globally sorted by code point, so lookups scan linearly for
// the covering range — mirroring the verified Lean find?. Each class is a
// partition, so at most one range covers a code point and the first match is
// the only match.

namespace UnicodeSecurity.Segmentation;

/// <summary>UAX #29 default extended grapheme cluster segmentation.</summary>
public static class Grapheme
{
    /// <summary>
    /// Grapheme_Cluster_Break class of <paramref name="cp"/>,
    /// <see cref="Gcb.Other"/> when uncovered.
    /// </summary>
    public static Gcb LookupGcb(uint cp)
    {
        foreach (var (first, last, cls) in GraphemeTables.GcbRanges)
        {
            if (first <= cp && cp <= last)
            {
                return cls;
            }
        }
        return Gcb.Other;
    }

    /// <summary>
    /// Indic_Conjunct_Break class of <paramref name="cp"/>,
    /// <see cref="Incb.None"/> when uncovered.
    /// </summary>
    public static Incb LookupIncb(uint cp)
    {
        foreach (var (first, last, cls) in GraphemeTables.IncbRanges)
        {
            if (first <= cp && cp <= last)
            {
                return cls;
            }
        }
        return Incb.None;
    }

    /// <summary>Whether <paramref name="cp"/> has the Extended_Pictographic property.</summary>
    public static bool IsExtPict(uint cp)
    {
        foreach (var (first, last) in GraphemeTables.ExtPictRanges)
        {
            if (first <= cp && cp <= last)
            {
                return true;
            }
        }
        return false;
    }

    /// <summary>GB11 left-context state: mirrors the Lean EPicState.</summary>
    private enum EpicState { None, AfterEp, AfterEpZwj }

    /// <summary>GB9c left-context state: mirrors the Lean InCBState.</summary>
    private enum IncbState { None, Consonant, Linker }

    /// <summary>Running scan state, mirroring the Lean State.</summary>
    private readonly struct State
    {
        public readonly Gcb? PrevClass;
        public readonly EpicState Epic;
        public readonly IncbState Incb;
        public readonly uint RiRun;

        public State(Gcb? prevClass, EpicState epic, IncbState incb, uint riRun)
        {
            PrevClass = prevClass;
            Epic = epic;
            Incb = incb;
            RiRun = riRun;
        }

        public static State Initial() => new(null, EpicState.None, IncbState.None, 0);
    }

    /// <summary>
    /// Whether a grapheme cluster break occurs immediately before
    /// <paramref name="cp"/> given the running state. Implements UAX #29
    /// GB1-GB999 in canonical order; first match wins, the trailing GB999
    /// breaks every otherwise-unmatched pair.
    /// </summary>
    private static bool ShouldBreakBefore(uint cp, in State s)
    {
        var bc = LookupGcb(cp);
        var incb = LookupIncb(cp);
        var isEp = IsExtPict(cp);
        if (s.PrevClass is not Gcb pc)
        {
            return true; // GB1: sot ÷
        }
        if (pc == Gcb.Cr && bc == Gcb.Lf)
        {
            return false; // GB3: CR × LF
        }
        if (pc == Gcb.Control || pc == Gcb.Cr || pc == Gcb.Lf)
        {
            return true; // GB4: (Control | CR | LF) ÷
        }
        if (bc == Gcb.Control || bc == Gcb.Cr || bc == Gcb.Lf)
        {
            return true; // GB5: ÷ (Control | CR | LF)
        }
        if (pc == Gcb.L && (bc == Gcb.L || bc == Gcb.V || bc == Gcb.Lv || bc == Gcb.Lvt))
        {
            return false; // GB6: L × (L | V | LV | LVT)
        }
        if ((pc == Gcb.Lv || pc == Gcb.V) && (bc == Gcb.V || bc == Gcb.T))
        {
            return false; // GB7: (LV | V) × (V | T)
        }
        if ((pc == Gcb.Lvt || pc == Gcb.T) && bc == Gcb.T)
        {
            return false; // GB8: (LVT | T) × T
        }
        if (bc == Gcb.Extend || bc == Gcb.Zwj)
        {
            return false; // GB9: × (Extend | ZWJ)
        }
        if (bc == Gcb.SpacingMark)
        {
            return false; // GB9a: × SpacingMark
        }
        if (pc == Gcb.Prepend)
        {
            return false; // GB9b: Prepend ×
        }
        if (s.Incb == IncbState.Linker && incb == Incb.Consonant)
        {
            return false; // GB9c: Consonant (Extend|Linker)* Linker (Extend|Linker)* × Consonant
        }
        if (s.Epic == EpicState.AfterEpZwj && isEp)
        {
            return false; // GB11: ExtPict Extend* ZWJ × ExtPict
        }
        if (bc == Gcb.RegionalIndicator && s.RiRun % 2 == 1)
        {
            return false; // GB12/GB13: odd-parity RI run extends
        }
        return true; // GB999: Any ÷ Any
    }

    /// <summary>Update the running state after consuming <paramref name="cp"/>. Mirrors the Lean advance.</summary>
    private static State Advance(uint cp, in State s)
    {
        var bc = LookupGcb(cp);
        var incb = LookupIncb(cp);
        var isEp = IsExtPict(cp);
        EpicState epic;
        if (isEp)
        {
            epic = EpicState.AfterEp;
        }
        else if (s.Epic == EpicState.AfterEp && bc == Gcb.Extend)
        {
            epic = EpicState.AfterEp;
        }
        else if (s.Epic == EpicState.AfterEp && bc == Gcb.Zwj)
        {
            epic = EpicState.AfterEpZwj;
        }
        else
        {
            epic = EpicState.None;
        }
        IncbState incbState;
        if (incb == Incb.Consonant)
        {
            incbState = IncbState.Consonant;
        }
        else if (s.Incb == IncbState.Consonant && incb == Incb.Linker)
        {
            incbState = IncbState.Linker;
        }
        else if (s.Incb == IncbState.Consonant && incb == Incb.Extend)
        {
            incbState = IncbState.Consonant;
        }
        else if (s.Incb == IncbState.Linker && incb == Incb.Linker)
        {
            incbState = IncbState.Linker;
        }
        else if (s.Incb == IncbState.Linker && incb == Incb.Extend)
        {
            incbState = IncbState.Linker;
        }
        else
        {
            incbState = IncbState.None;
        }
        var riRun = bc == Gcb.RegionalIndicator ? s.RiRun + 1 : 0;
        return new State(bc, epic, incbState, riRun);
    }

    /// <summary>
    /// Boundary mask of length <c>cps.Count + 1</c>. Entry <c>i</c> is
    /// <c>true</c> when a grapheme cluster break occurs immediately before
    /// position <c>i</c> — entry <c>0</c> is the GB1 start-of-text break, entry
    /// <c>cps.Count</c> the GB2 end-of-text break, both always <c>true</c>.
    /// Mirrors the Lean graphemeBreaks.
    /// </summary>
    public static bool[] GraphemeBreaks(IReadOnlyList<int> cps)
    {
        var bs = new bool[cps.Count + 1];
        var s = State.Initial();
        for (var i = 0; i < cps.Count; i++)
        {
            var cp = (uint)cps[i];
            bs[i] = ShouldBreakBefore(cp, s);
            s = Advance(cp, s);
        }
        bs[cps.Count] = true; // GB2: eot ÷
        return bs;
    }

    /// <summary>
    /// Split <paramref name="cps"/> into grapheme clusters (the code points
    /// between consecutive boundaries).
    /// </summary>
    public static List<List<int>> GraphemeClusters(IReadOnlyList<int> cps)
    {
        var breaks = GraphemeBreaks(cps);
        var outClusters = new List<List<int>>();
        var cur = new List<int>();
        for (var i = 0; i < cps.Count; i++)
        {
            if (breaks[i] && cur.Count > 0)
            {
                outClusters.Add(cur);
                cur = new List<int>();
            }
            cur.Add(cps[i]);
        }
        if (cur.Count > 0)
        {
            outClusters.Add(cur);
        }
        return outClusters;
    }
}
